text/0005-white-flag/0005-white-flag.md

```diff
@@ -0,0 +1,155 @@
+conflict.
+
+## Deterministically ordering the Tangle
+
+When a new milestone is broadcasted to the network, nodes will need to order the set of bundles it confirms.
+
+A subset of the Tangle can be ordered depending on many of its properties (e.g. alphanumeric sort of the bundle hashes);
+however, to compute the ledger state, a graph traversal has to be done so it can be used to order the bundles in a
+deterministic order with no extra overhead.
+
+This ordering is then defined as a [topological ordering](https://en.wikipedia.org/wiki/Topological_sorting) because
+it respects the dependency of bundles, ensuring that the `trunk` and `branch` of a bundle are applied before it. Since
+there are multiple valid topological orders for the same graph and in order to avoid conflicting ledger states it is
+required that all nodes apply bundles in the exact same order.
+
+For this reason, this RFC propose an order that has to be rigorously followed by all node implementations. This order is
+the topological ordering generated by a post-order [Depth-First Search (DFS)](https://en.wikipedia.org/wiki/Depth-first_search)
+starting from a milestone and by going first through `trunk` bundle, then `branch` bundle and finally current bundles.
+Since only a subset of bundles is considered, the stopping condition of this DFS is reaching bundles that are already
+confirmed by another milestone.
+
+## Applying first bundle(s) that does not violate the ledger state
+
+If a conflict is occurring in the set of bundles confirmed by a milestone, nodes have to apply the first - with regards
+to the order previously proposed - of the conflicting bundles to the ledger and ignore all the others.
+
+Once a bundle is marked as ignored, this is final and can't be changed by a later milestone.
+
+Since the ledger state is maintained from one milestone to another, a bundle conflicting with a bundle already confirmed
+by a previous milestone would also obviously be ignored.
+
+## Pseudo-code
+
+The following algorithm describes the process of updating the ledger state which is usually triggered by the arrival of
+a new milestone confirming many new bundles.
+
+Even though the tangle is a graph made of transactions, for the sake of this pseudo-code it is considered as a graph of
+bundles. For this reason, when the `trunk` and/or the `branch` of a bundle are mentioned, they are actually referring to
+the `trunk` and/or `branch` of the last transaction of the bundle.
+
+This algorithm is to be followed with care when implementing a node to avoid differences in the ledger state.
+
+```
+update_ledger_state(ledger, milestone) {
+    s = new Stack()
+    visited = new Set()
+
+    s.push(milestone)
+
+    while (!s.empty()) {
+        curr = s.peek()
+
+        // only apply if you reached a leaf or both approvees have been visited or confirmed by another milestone
+        if ((curr.trunk == null || visited.contains(curr.trunk) || curr.trunk.snapshot_index != milestone.index)
```

Comment:
I just realized that `curr.trunk == null` is a mistake

Because if it is null it may mean that we have a non-solid subtangle...
What the check here should be is if `curr.trunk == genesis`

Theoretically, `curr.trunk` should never be null because we solidify the milestone before applying it...
Maybe we should point this out in the RFC?

Reply:
if it is null then there should be an error I believe, because we have holes in the tangle

Reply:
Sure, I even wondered if it should be an SEP check instead of a genesis check but let's not over complicate it.

Reply:
well yes actually
SEP is like genesis